### DISCUSSION TITLE
Diffusion term in k transport equation

### DISCUSSION THREAD

**Discussion by JuanSembla** [su2code/SU2#1837] [Q&A]
Hello all! I hope you're doing well.

I'm working on turbulence modelling for compressible flows. More precisely, the idea is to obtain data driven models that work as corrections for Menter's k-w SST and so, I'll be modifying the k and omega equations. In order to do so, I would need to access the diffusion term that goes like $$\partial / \partial x_j [(\nu + \sigma_k * \nu_t) \partial k / \partial x_j]$$. I've seen that in CTurbSSTSolver.cpp, from line 319 onwards, the code goes like:

/*--- Compute the source term ---*/
    
auto residual = numerics->ComputeResidual(config);

/*--- Subtract residual and the Jacobian ---*/

LinSysRes.SubtractBlock(iPoint, residual);
if (implicit) Jacobian.SubtractBlock2Diag(iPoint, residual.jacobian_i);

Is the diffusion term allocated in residual.jacobian_i? And, if it is, where does this allocation take place? I've gone through turb_diffusion.hpp and scalar_diffusion.hpp but I don't seem to have found it yet. 

Thank you in advance for your time and cooperation, I hope you have a nice day.

Regards,

JuanSembla.

**Comment by bigfooted** [comment]
You need to be here for the actual diffusion term that you mention (around line 230):
https://github.com/su2code/SU2/blob/master/SU2_CFD/include/numerics/turbulent/turb_diffusion.hpp
and finishresidualcalc is then called in ComputeResidual from here:
https://github.com/su2code/SU2/blob/master/SU2_CFD/include/numerics/scalar/scalar_diffusion.hpp
which is called by all scalar solvers (turbulence and species transport)

The residual is then added to the linear system, and its Jacobian is added to the Jacobian, see the theory, especially eq. (49) here:
https://su2code.github.io/documents/AIAA-2018-3111.pdf

**Reply by JuanSembla** [reply]
Thanks a lot for your answer, it was really helpful. I believe I understand better how the code works but I still have some things missing. Indeed, regarding the flux that is computed in finishresidualcalc in scalar_diffusion.hpp, how does he ends up being the residual that is computed when doing:

auto residual = numerics->ComputeResidual(config); (around line 120 of CScalarSolver.hpp) ?

As far as I understand, this residual is the actual diffusion term and should be the divergence of the flux that is previously computed by finishresidualcalc. Is this correct? How does SU2 works his way to compute this divergence and then obtain the residual.

Thanks in advance and have a good day!

JuanSembla.

**Reply by pcarruscag** [reply]
That is somewhat of a naming problem, the numerics classes compute fluxes, the solver then sums those fluxes over the faces of each control volume (the sum is the residual).

**Reply by JuanSembla** [reply]
Ok, thanks. And where does this sum take place?
Thanks for your help.

**Reply by pcarruscag** [reply]
Everywhere you see LinSysRes.SubtractBlock(iPoint, residual); (or AddBlock).
So mainly in Upwind_Residual (and Viscous_Residual) and then on boundary conditions
